command/command.go

```diff
@@ -1,129 +1,113 @@
 	"os/exec"
 	"strconv"
 	"strings"
 )
 
-// ----------
-
-// Model ...
-type Model struct {
-	cmd *exec.Cmd
+// Opts ...
+type Opts struct {
+	Stdout io.Writer
+	Stderr io.Writer
+	Env    []string
+	Dir    string
 }
 
-// New ...
-func New(name string, args ...string) *Model {
-	return &Model{
-		cmd: exec.Command(name, args...),
-	}
+// Factory ...
+type Factory interface {
+	Create(name string, args []string, opts *Opts) Command
 }
 
-// NewWithStandardOuts - same as NewCommand, but sets the command's
-// stdout and stderr to the standard (OS) out (os.Stdout) and err (os.Stderr)
-func NewWithStandardOuts(name string, args ...string) *Model {
-	return New(name, args...).SetStdout(os.Stdout).SetStderr(os.Stderr)
+type defaultFactory struct {
+	envRepository env.Repository
 }
 
-// NewWithParams ...
-func NewWithParams(params ...string) (*Model, error) {
-	if len(params) == 0 {
-		return nil, errors.New("no command provided")
-	} else if len(params) == 1 {
-		return New(params[0]), nil
-	}
-
-	return New(params[0], params[1:]...), nil
+// NewFactory ...
+func NewFactory(envRepository env.Repository) Factory {
+	return defaultFactory{envRepository: envRepository}
 }
 
-// NewFromSlice ...
-func NewFromSlice(slice []string) (*Model, error) {
-	return NewWithParams(slice...)
-}
-
-// NewWithCmd ...
-func NewWithCmd(cmd *exec.Cmd) *Model {
-	return &Model{
-		cmd: cmd,
+// Create ...
+func (d defaultFactory) Create(name string, args []string, opts *Opts) Command {
+	cmd := exec.Command(name, args...)
+	if opts != nil {
+		cmd.Stdout = opts.Stdout
+		cmd.Stderr = opts.Stderr
+
+		// If Env is nil, the new process uses the current process's
+		// environment.
+		// If we pass env vars we want to append them to the
+		// current process's environment.
+		cmd.Env = append(d.envRepository.List(), opts.Env...)
+		cmd.Dir = opts.Dir
 	}
+	return defaultCommand{cmd}
 }
 
-// GetCmd ...
-func (m *Model) GetCmd() *exec.Cmd {
-	return m.cmd
-}
-
-// SetDir ...
-func (m *Model) SetDir(dir string) *Model {
-	m.cmd.Dir = dir
-	return m
+// Command ...
+type Command interface {
+	PrintableCommandArgs() string
+	Run() error
+	RunAndReturnExitCode() (int, error)
+	RunAndReturnTrimmedOutput() (string, error)
+	RunAndReturnTrimmedCombinedOutput() (string, error)
 }
 
-// SetEnvs ...
-func (m *Model) SetEnvs(envs ...string) *Model {
-	m.cmd.Env = envs
-	return m
-}
-
-// AppendEnvs - appends the envs to the current os.Environ()
-// Calling this multiple times will NOT appens the envs one by one,
-// only the last "envs" set will be appended to os.Environ()!
-func (m *Model) AppendEnvs(envs ...string) *Model {
-	return m.SetEnvs(append(os.Environ(), envs...)...)
-}
-
-// SetStdin ...
-func (m *Model) SetStdin(in io.Reader) *Model {
-	m.cmd.Stdin = in
-	return m
+type defaultCommand struct {
+	cmd *exec.Cmd
 }
 
-// SetStdout ...
-func (m *Model) SetStdout(out io.Writer) *Model {
-	m.cmd.Stdout = out
-	return m
+// NewCommand ...
+func NewCommand(cmd *exec.Cmd) Command {
```

Comment:
What's the use case for this constructor?

Reply:
It is used in the test right now, but I agree that maybe in real world we don't want to wrap standard commands to our

Reply:
It may be useful to wrap a go command in the new one to avoid a big refactor, in the bitrise CLI repo for example I remember doing something  that.

Reply:
Just do the refactor :) We broke the entire command package anyway

Reply:
:D I must agree you are right.